ETHCHAT.md

```diff
@@ -0,0 +1,230 @@
+2.  Execution of X3DH: This step will output a common secret key SK together with an additional data vector AD. Both will be used in the Double Ratchet algorithm initialization.
+3.  Execution of the Double Ratchet algorithm for forward secure, authenticated communications, using the common secret key SK, obtained from X3DH, as a root key.
+
+## Cryptographic functions required
+-   XEdDSA for digital signatures involved in the X3DH key generation.
+-   SHA256 for hashing and the generation of HMACs.
+-   AES256-CBC for the encryption/decryption of messages.
+
+## Considerations on the X3DH initialization
+This scheme requires working on specific elliptic curves which differ from those used by Ethereum. 
+To be precise, Ethereum makes use of the curve secp256k1, whereas X3DH requires either X25519 or X448. For security reasons one must work on the curve X448.
+
+Bob and Alice must define a key pair (ik, IK) where:
+-   The key ik must be kept secret,
+-   and the key IK is public.
+
+Bob will not be able to use his Ethereum public key during this stage due to incompatibilities with the involved elliptic curves, therefore it will be required to generate new keys. 
+This can be done using the basepoint $G$ for X448 and $ik \in \mathbb{Z}_p$ a random integer:
+
+$$ IK = ik \cdot G $$
+
+The scheme X3DH will also require the generation of a public key SPK which will be generated repeating the above process: one takes $spk \in \mathbb{Z}_p$ a secret random integer and computes:
+
+$$ SPK = spk \cdot G $$
+
+SPK is a public key generated and stored at medium-term. 
+It is called a signed prekey because Bob also needs to store a public key certificate of SPK using IK. 
+Both signed prekey and the certificate must undergo periodic replacement, 
+a process that entails the generation of a fresh signed prekey. 
+After replacing the key, 
+Bob keeps the old private key of SPK for some interval, dependant on the implementation.
+This allows Bob to decrypt delayed messages. 
+It is important that Bob does not reuse SPKs. 
+This action is pivotal for ensuring forward secrecy, as these keys are integral for recalculating the shared secret employed in decrypting historical messages.
+
+It will be required to sign SPK for authentication. Following the specification of X3DH, one will use the digital signature scheme XEdDSA and define:
+
+$$ SigSPK = XEdDSA(ik, Encode(SPK)) $$
+
+A final step requires the definition of a  _prekey bundle_  given by the tuple
+
+$$ prekey\_bundle = (IK, SPK, SigSPK, \{OPK_i\}_i) $$
+
+Where the different one-time keys OPK are points in X448 generated from a random integer $opk \in \mathbb{Z}_p$ and computed by performing
+
+$$ OPK = opk\cdot G $$
+
+Before sending an initial message to Bob, Alice will generate an AD vector as described in the documentation:
+
+$$ AD = Encode(IK_A)|| Encode(IK_B) $$
```

Comment:
Should we specify what is the Encoding function?

Reply:
Comment added